Flush classloader cache

eatkins opened the pull request:
I am trying to clean up as many of the URLClassLoaders created by sbt as possible. The loaders managed by a ClassLoaderCache cannot presently be closed, even when we are sure they won't be used again.

eatkins committed:
Add ability to flush ClassLoaderCache

Even though the ClassLoaderCache uses soft references, it is still
holding references to URLClassLoaders, which, in turn, may be holding
open handles to files. We should have a mechanism for explicitly closing
these.

eatkins committed:
Remove comment

The ClassLoaderCache is used in may places so it is clearly cannot just
be considered a "Hack for testing only".

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=develop&zinc=pull/674/head&io=develop&librarymanagement=develop&util=develop&website=develop&deploy_to_bintray=false&run_sbt_community=false&verbose_dbuild=false&community_verbose=false">this link.</A>

eatkins committed:
Make ClassLoaderCache extendable by delegation

We are having a lot of problems in sbt 1.3.0-RC1 with metaspace
utilization. Leaking classlaoders in the ClassLoaderCache is a primary
culprit. I want to be able to modify the ClassLoaderCache implementation
in sbt so that we can manually close it in a number of situations (such
as when the user runs '++' directly or indirectly. To do this, I define
a trait, AbstractClassLoaderCache that specifies the public api of
ClassLoaderCache and then refactor ClassLoaderCache so that the primary
constructor takes an instance of AbstractClassLoaderCache and delegates
to it. Finally, I provide a default implementation that is identical to
the old version.

eed3si9n commented:
Is this conflicting with the Scalafmt change?

eatkins commented:
I have no idea how I managed to screw up the branch so badly. It was a rebase gone horribly awry. Anyway, fixed now.
